File: backend/Services/ProjectService.cs
using backend.Models;

namespace backend.Services;

public class ProjectService
{
    private readonly List<Project> _projects = [];
    private int _nextProjectId = 1;
    private int _nextTaskId = 1;

    public IEnumerable<Project> GetAll() => _projects;
    public Project? Get(int id) => _projects.FirstOrDefault(p => p.Id == id);
    public Project Add(string name)
    {
        var project = new Project { Id = _nextProjectId++, Name = name};
        _projects.Add(project);
        return project;
    }
    public TaskItem? AddTask(int projectId, string title)
    {
        var project = Get(projectId);
        if (project is null) return null;

        var task = new TaskItem { Id = _nextTaskId++, Title = title, IsCompleted = false};
        project.Tasks.Add(task);
        return task;
    }
    
    public bool ToggleTask(int projectId, int taskId)
    {
        var project = Get(projectId);
        var task = project?.Tasks.FirstOrDefault(t => t.Id == taskId);
        if (task == null) return false;

        task.IsCompleted = !task.IsCompleted;
        return true;
    }
}

File: frontend/src/app/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Project, TaskItem } from '../models/project';

@Injectable({
  providedIn: 'root',
})
export class ProjectService {
  private apiUrl = 'http://localhost:5134/api/projects';

  constructor(private http: HttpClient) {}

  // Get all projects
  getProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(this.apiUrl);
  }

  // Get single project
  getProject(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }

  // Create new project
  createProject(name: string): Observable<Project> {
    return this.http.post<Project>(
      this.apiUrl,
      { name },
      {
        headers: { 'Content-Type': 'application/json' },
      }
    );
  }

  // Add task
  addTask(projectId: number, title: string): Observable<TaskItem> {
    return this.http.post<TaskItem>(
      `${this.apiUrl}/${projectId}/tasks`,
      { title },
      { headers: { 'Content-Type': 'application/json' } }
    );
  }

  // Toggle task complete
  toggleTask(projectId: number, taskId: number): Observable<void> {
    return this.http.put<void>(
      `${this.apiUrl}/${projectId}/tasks/${taskId}/toggle`,
      {}
    );
  }
}
